src/hotspot/share/jvmci/jvmciRuntime.hpp

```diff
@@ -124,24 +125,63 @@ class JVMCIRuntime: public CHeapObj<mtJVMCI> {
   // Initialization state of this JVMCIRuntime.
   InitState _init_state;
 
+  // Initialization state of the references to classes, methods
+  // and fields in the JVMCI shared library.
+  static InitState _shared_library_javavm_refs_init_state;
+
+  // Initialization state of the references to classes, methods
+  // and fields in HotSpot metadata.
+  static InitState _hotspot_javavm_refs_init_state;
+
   // A wrapper for a VM scoped JNI global handle (i.e. JVMCIEnv::make_global)
   // to a HotSpotJVMCIRuntime instance. This JNI global handle must never
   // be explicitly destroyed as it can be accessed in a racy way during
   // JVMCI shutdown. Furthermore, it will be reclaimed when
   // the VM or shared library JavaVM managing the handle dies.
   JVMCIObject _HotSpotJVMCIRuntime_instance;
 
+  // Lock for operations that may be performed by
+  // any thread attached this runtime. To avoid deadlock,
+  // this lock must always be acquired before JVMCI_lock.
+  Monitor* _lock;
+
   // Result of calling JNI_CreateJavaVM in the JVMCI shared library.
-  // Must only be modified under JVMCI_lock.
+  // Must only be mutated under _lock.
   volatile JavaVM* _shared_library_javavm;
 
+  // Id for _shared_library_javavm.
+  int _shared_library_javavm_id;
+
+  // Position and link in global list of JVMCI shared library runtimes.
   // The HotSpot heap based runtime will have an id of -1 and the
-  // JVMCI shared library runtime will have an id of 0.
+  // runtime reserved for threads attaching during JVMCI shutdown
+  // will have an id of -2.
   int _id;
+  JVMCIRuntime* _next;
 
   // Handles to Metadata objects.
   MetadataHandles* _metadata_handles;
 
+  // List of jobjects allocated via make_global. This is to support
+  // destroying remaining JNI handles when the JavaVM associated
+  // with this runtime is shutdown.
+  GrowableArray<jobject>* _jobjects;
```

Comment:
Interesting proposal. What would be the advantage of `OopHandle` over `jobject`? That require a bunch of `reinterpret_cast`ing between `jobject` and `OopHandle` which I'm not sure is even possible?

Reply:
The suggestion was about managing the handles and associated values (and I'm not even sure it would help there).  But maybe `jobject` is too deeply embedded in JVMCI usage?  Or maybe there are places where it really needs to be using `jobject` rather than `oop*` or a thin wrapper over `oop*` like `OopHandle`?

Reply:
I'm avoiding `jobject` altogether now: https://github.com/openjdk/jdk/pull/8262/commits/fc9dd87fe87271c33f9861e97c84d1cc681e40c6